[py]Unexpected NULL When Reading Nested ROW type with Nullable Fields

The following SQL:
```
CREATE TABLE tbl(row_row ROW(v1 ROW(v11 VARCHAR NULL), v2 ROW(v21 VARCHAR NULL)));

CREATE MATERIALIZED VIEW v AS SELECT *  FROM tbl;
```
Given input: 
`"row_row": {"v1": {"v11": "hi"}, "v2": {"v21": "ola!!"}}`

Produces the following output:
**[{'row_row': {'v1': {'v11': 'hi'}, 'v2': {'v21': None}}}]**

### OBSERVATIONS
1. **Removing NULL from v21**: 
 `CREATE TABLE tbl(row_row ROW(v1 ROW(v11 VARCHAR NULL), v2 ROW(v21 VARCHAR)));`
Correctly returns: **[{'row_row': {'v1': {'v11': 'hi'}, 'v2': {'v21': 'ola!!'}}}]**
2. **Removing NULL from both v11 and v21**: 
`CREATE TABLE tbl(row_row ROW(v1 ROW(v11 VARCHAR), v2 ROW(v21 VARCHAR)));`
Produces the following parse error: 
**E               Parse error (event #1): failed to deserialize JSON record: error parsing field 'row_row': {"field":"v2","description":"missing field `v11` at line 1 column 56"} at line 1 column 57
E           Invalid fragment: '{"row_row": {"v1": {"v11": "hi"}, "v2": {"v21": "ola!!"}}}'**
3. **Using different types for v11 and v21**:
 `CREATE TABLE tbl(row_row ROW(v1 ROW(v11 INT NULL), v2 ROW(v21 VARCHAR NULL)));`
Given that the input for **'v11' = 12**, correctly returns: **[{'row_row': {'v1': {'v11': 12}, 'v2': {'v21': 'ola!!'}}}]**
4. **Removing NULL for v11 and v21 with different types**: 
 `CREATE TABLE tbl(row_row ROW(v1 ROW(v11 INT), v2 ROW(v21 VARCHAR)));`
Also correctly returns: **[{'row_row': {'v1': {'v11': 12}, 'v2': {'v21': 'ola!!'}}}]**

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/DBSPCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/DBSPCompiler.java
@@ -123,7 +123,7 @@ public class DBSPCompiler implements IWritesLogs, ICompilerComponent, IErrorRepo
         builder.append(typeName);
         builder.append("[");
         for (var field: fields) {
-            builder.append(field.getSanitizedName())
+            builder.append(field.getName())
                     .append(":")
                     .append(field.getType())
                     .append(",");

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/CatalogTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/CatalogTests.java
@@ -7,14 +7,20 @@ import org.dbsp.sqlCompiler.compiler.DBSPCompiler;
 import org.dbsp.sqlCompiler.compiler.frontend.calciteCompiler.ProgramIdentifier;
 import org.dbsp.sqlCompiler.compiler.frontend.calciteObject.CalciteObject;
 import org.dbsp.sqlCompiler.compiler.sql.tools.BaseSQLTests;
+import org.dbsp.sqlCompiler.compiler.visitors.inner.InnerVisitor;
+import org.dbsp.sqlCompiler.ir.IDBSPInnerNode;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
+import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeStruct;
 import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeTuple;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeInteger;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeString;
 import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeArray;
 import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeZSet;
+import org.dbsp.util.Linq;
 import org.junit.Assert;
 import org.junit.Test;
+
+import java.util.List;
 
 /** Tests that emit Rust code using the catalog. */
 public class CatalogTests extends BaseSQLTests {
@@ -25,6 +31,27 @@ public class CatalogTests extends BaseSQLTests {
         result.ioOptions.emitHandles = false;
         result.languageOptions.unrestrictedIOTypes = false;
         return result;
+    }
+
+    @Test
+    public void issue3902() {
+        var ccs = this.getCCS("""
+                CREATE TABLE tbl(row_row ROW(v1 ROW(v11 VARCHAR NULL), v2 ROW(v21 VARCHAR NULL)));
+                CREATE MATERIALIZED VIEW v AS SELECT *  FROM tbl;""");
+        InnerVisitor checkStruct = new InnerVisitor(ccs.compiler) {
+            @Override
+            public void postorder(DBSPTypeStruct type) {
+                if (type.fields.size() == 2) {
+                    // Before the bug was fixes the two fields had the same type
+                    List<ProgramIdentifier> names = Linq.list(type.getFieldNames());
+                    assert names.size() == 2;
+                    DBSPType f0 = type.getFieldType(names.get(0));
+                    DBSPType f1 = type.getFieldType(names.get(1));
+                    Assert.assertFalse(f0.sameType(f1));
+                }
+            }
+        };
+        ccs.visit(checkStruct.getCircuitVisitor(false));
     }
 
     @Test

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
@@ -26,6 +26,11 @@ import org.junit.Test;
 import java.util.HashSet;
 
 public class RegressionTests extends SqlIoTest {
+    @Test @Ignore("https://issues.apache.org/jira/browse/CALCITE-6976")
+    public void issue3913() {
+        this.getCC("CREATE VIEW V AS SELECT SUBSTR(TIMESTAMP '2024-01-01 10:00:00', 10)");
+    }
+
     @Test
     public void issue3653() {
         this.statementsFailingInCompilation("CREATE TYPE t AS (related t ARRAY);",

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/resources/metadataTests-generateDF.json
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/resources/metadataTests-generateDF.json
@@ -94,7 +94,7 @@
         "final": 1
       },
       "positions": [],
-      "persistent_id": "c34714d45e88dd13c0abe1a709eee661cd6494ff2736f095c24b5add3103cde8"
+      "persistent_id": "ab7ba64d3de62f8105f8d9ad92b71271274bc6e04fd17c459ce7bec4ddcb869f"
     }, "s3": {
       "operation": "map_index",
       "inputs": [
@@ -109,7 +109,7 @@
           }]
       },
       "positions": [],
-      "persistent_id": "90fc2d18f514fb1195d45924101472c0f6edd8ae7bd8b48c297f7c14e4559314"
+      "persistent_id": "1e7ec9af3140a9c03cd72189f54e5173d8e2898db245947ec5fe63fcd57f01d0"
     }, "s4": {
       "operation": "differentiate",
       "inputs": [
@@ -119,7 +119,7 @@
         "partial": 3
       },
       "positions": [],
-      "persistent_id": "ea0f5e09135c0295b5bab326b5904ac8c8c3907080db811a3057011e93b14858"
+      "persistent_id": "7ff9e7087aac79f5694df3b30ca0f417f387862e78c97c518bf5d8919d04f2a4"
     }, "s5": {
       "operation": "aggregate_linear_postprocess",
       "inputs": [
@@ -131,7 +131,7 @@
       "positions": [
         {"start_line_number":2,"start_column":25,"end_line_number":2,"end_column":33}
       ],
-      "persistent_id": "0eb727d4821dead2e58bcfd0375ec4f8ad1a79e73d83b5162ee222522384f4bc"
+      "persistent_id": "acf4333e87c8a9c0ee6a52dbc14b59ade23dcf57589c01056a35a23bea4bbe28"
     }, "s6": {
       "operation": "map",
       "inputs": [
@@ -141,7 +141,7 @@
         "partial": 3
       },
       "positions": [],
-      "persistent_id": "91350935cf25c4df4b2cd694cefe5cde71ff21c1045adf6d706444ae5632da33"
+      "persistent_id": "1411be44836dd4fa829a3ced8abd64b9137bf43e65350ec03053f67e04396437"
     }, "s7": {
       "operation": "map",
       "inputs": [
@@ -151,7 +151,7 @@
         "partial": 3
       },
       "positions": [],
-      "persistent_id": "14acabbb0e718fd811810bd0dd2eb481977cd8b7af68c38a101fc29a58cb49ba"
+      "persistent_id": "d5cdd7219fb8d20e372821140e915df906b5759012c73961d68235a11493b48d"
     }, "s8": {
       "operation": "neg",
       "inputs": [
@@ -161,7 +161,7 @@
         "partial": 3
       },
       "positions": [],
-      "persistent_id": "fd3f34052e023ad8f2278cc105d458641754c726088d6020c6a380eeb17c3487"
+      "persistent_id": "a9a52124256e347f4fe90ccaa7ea975b84999d408c32107b6694816fa19db793"
     }, "s9": {
       "operation": "integrate",
       "inputs": [
@@ -171,7 +171,7 @@
         "partial": 3
       },
       "positions": [],
-      "persistent_id": "149aff585cd89bedb3e340c385788cd0d72b2c5f5b2524ea19e9cde2c49916f3"
+      "persistent_id": "43585b3a8427060c16490f77964e3ec7606469004ea00f59b2769b584c498211"
     }, "s10": {
       "operation": "integrate",
       "inputs": [
@@ -181,7 +181,7 @@
         "partial": 3
       },
       "positions": [],
-      "persistent_id": "cb5603d163f0e6fd30330e7f9973d00f91cac1c9d0739c7be7735e92601d3e53"
+      "persistent_id": "6bb9b96f3d58ee381084d28fd16d09e183ce304bcd6f2d6695bc9129f861c9f9"
     }, "s11": {
       "operation": "sum",
       "inputs": [
@@ -193,7 +193,7 @@
         "partial": 3
       },
       "positions": [],
-      "persistent_id": "45a4e6bee369f75351d512573d1e54b9717a5c64d9bf62e8ccbb6d238ce497ec"
+      "persistent_id": "d838e77e8e11928839dec855d5a03330e556d13b0e9e8a901af8259ffd5bab1c"
     }, "s12": {
       "operation": "inspect",
       "inputs": [
@@ -203,7 +203,7 @@
         "final": 3
       },
       "positions": [],
-      "persistent_id": "fa20d104fa4772b6f6521bdf29331c01ddc49cc4b7f9d7afc441298b2e95ed34"
+      "persistent_id": "6ade4797370fb9bb7fcfd441733a88e0f5cd6b64815a3c0e0dce0aee3ef839d0"
     }, "s13": {
       "operation": "inspect",
       "inputs": [
@@ -213,7 +213,7 @@
         "final": 0
       },
       "positions": [],
-      "persistent_id": "6f993c13fc9c17553263ce8f613ccd7335ef057c1d05214ea4fdc4f6a1218793"
+      "persistent_id": "933ebf782e1fe804fe85c4d0f3688bdb5234b386c2834892776e692acd9781d9"
     }
   },
   "sources": "CREATE TABLE T (COL1 INT NOT NULL, COL2 DOUBLE NOT NULL);\nCREATE VIEW V AS SELECT SUM(COL1) FROM T;\n"

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/resources/metadataTests-generateDFRecursive.json
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/resources/metadataTests-generateDFRecursive.json
@@ -289,7 +289,7 @@
           "final": 2
         },
         "positions": [],
-        "persistent_id": "78c73d7673ad511b1472b1f145df83c836ab6f3963b7d61f873d8217e0374610"
+        "persistent_id": "f760e73e7ed394ea9ce4a311c3b97958ab15a13bb2f360a259e45f6984ff7ac5"
       },
       "s4": {
         "operation": "flat_map_index",
@@ -308,7 +308,7 @@
           {"start_line_number":1,"start_column":1,"end_line_number":1,"end_column":50},
           {"start_line_number":20,"start_column":27,"end_line_number":20,"end_column":37}
         ],
-        "persistent_id": "a05d1bb2948240003bb8d8bc03c31fa7427e559b587c47d9633d3f65016b56a2"
+        "persistent_id": "c1c5233542d6daeb994aa33dcf954af8a5df15c2e5879b4d9e52a64c557fa144"
       },
       "s5": {
         "operation": "flat_map_index",
@@ -330,7 +330,7 @@
           {"start_line_number":19,"start_column":17,"end_line_number":19,"end_column":26},
           {"start_line_number":20,"start_column":11,"end_line_number":20,"end_column":21}
         ],
-        "persistent_id": "cf57508bf361b88abcf8f94f2f9f7f3439fe69d013abb32d8a778508059d5937"
+        "persistent_id": "83809e87c5f3080f95a60dccd59210486fabab0d0cff6f6212bd628a8281bda4"
       },
       "s6": {
         "operation": "delta0",
@@ -369,7 +369,7 @@
           {"start_line_number":16,"start_column":9,"end_line_number":16,"end_column":33},
           {"start_line_number":15,"start_column":9,"end_line_number":15,"end_column":18}
         ],
-        "persistent_id": "ee036b6dcd0e5861401629b33db28974f61cebef835167fa623b7423912c5e18"
+        "persistent_id": "87e06cf7a2cafb547f1832b80cda3a373d93a74ba60797ea2ef78d4d1cd24b68"
       },
       "s9": {
         "operation": "sum",
@@ -381,7 +381,7 @@
           "final": 8
         },
         "positions": [],
-        "persistent_id": "5f95174f92c888c6d419209df06d0a3a787a9b4b46475bec95f68bd0e69dec4d"
+        "persistent_id": "662552a13949639939a6cb7b9a34e1b99a5e106f65544446cd3bda636f517f27"
       }
     }, "s10": {
       "operation": "inspect",
@@ -392,7 +392,7 @@
         "final": 8
       },
       "positions": [],
-      "persistent_id": "6f527753e537a606f7cfedd98c33037c3a3d14b8574c5dfffa21329863925147"
+      "persistent_id": "79e9e16c49264531840e6128f9bb18cb9633098f56199c923c7083a93508fe57"
     }, "s11": {
       "operation": "inspect",
       "inputs": [
@@ -402,7 +402,7 @@
         "final": 0
       },
       "positions": [],
-      "persistent_id": "6f993c13fc9c17553263ce8f613ccd7335ef057c1d05214ea4fdc4f6a1218793"
+      "persistent_id": "933ebf782e1fe804fe85c4d0f3688bdb5234b386c2834892776e692acd9781d9"
     }
   },
   "sources": "DECLARE RECURSIVE view fibonacci(n INT, value INT);\ncreate table input (x int);\n\ncreate view fibonacci AS\n(\n    -- Base case: first two Fibonacci numbers\n    select 0 as n, 0 as value\n    union all\n    select 1 as n, 1 as value\n)\nunion all\n(\n    -- Compute F(n)=F(n-1)+F(n-2)\n    select\n        prev.n + 1 as n,\n        (prev.value + curr.value) as value\n    from fibonacci as curr\n    join fibonacci as prev\n    on prev.n = curr.n - 1\n    where curr.n < 10 and prev.n < 10\n);\n"

